wallet/src/service/Offers.ts

```diff
@@ -207,45 +208,95 @@ export const getOfferService = (
     return signSpendAction(offer.spendAction);
   };
 
-  const cancelOffer = _id => {
-    console.log('TODO: cancel offer');
+  const tryExitOffer = async (id: number) => {
+    const action = await E(boardIdMarshaller).serialize(
+      harden({
+        method: 'tryExitOffer',
+        offerId: id,
+      }),
+    );
+
+    return signSpendAction(JSON.stringify(action));
   };
 
-  const watchChainOffers = async () => {
+  const watchOfferUpdates = async () => {
     for await (const status of makeAsyncIterableFromNotifier(
-      chainOffersNotifier,
+      offerUpdatesNotifier,
     )) {
       console.log('offerStatus', { status, offers });
       const id = status && Number(status?.id);
       const oldOffer = offers.get(id);
       if (!oldOffer) {
         console.warn('Update for unknown offer, doing nothing.');
       } else {
-        if (status.error !== undefined) {
+        if (status.numWantsSatisfied !== undefined) {
           offers.set(id, {
             ...oldOffer,
             id,
-            status: OfferUIStatus.rejected,
-            error: `${status.error}`,
+            status:
+              status.numWantsSatisfied === 0
+                ? OfferUIStatus.refunded
+                : OfferUIStatus.accepted,
+            isSeated: false,
+            error: status.error && `${status.error}`,
           });
           remove(smartWalletKey, id);
-        } else if (status.numWantsSatisfied !== undefined) {
+        } else if (status.error !== undefined) {
           offers.set(id, {
             ...oldOffer,
             id,
-            status: OfferUIStatus.accepted,
+            status: OfferUIStatus.rejected,
+            error: `${status.error}`,
           });
-          remove(smartWalletKey, id);
-        } else if (status.numWantsSatisfied === undefined) {
+          if (!oldOffer.isSeated) {
+            remove(smartWalletKey, id);
+          }
+        }
+        broadcastUpdates();
+      }
+    }
+  };
+
+  const watchPendingOffers = async (brandToPurse: Map<Brand, PurseInfo>) => {
+    const makeProposalEntriesDisplayable = (proposalEntries: {
+      [key: string]: Amount;
+    }) =>
+      Object.fromEntries(
+        Object.entries(proposalEntries).map(([kw, entry]) => {
+          const pursePetname = brandToPurse.get(entry.brand)?.pursePetname;
+          const value = String(entry.value);
+          return [kw, { pursePetname, value }];
+        }),
+      );
+
+    for await (const pendingOffers of makeAsyncIterableFromNotifier(
+      pendingOffersNotifier,
+    )) {
+      console.log('pending offers', pendingOffers);
+      pendingOffers?.forEach(([_, o]) => {
```

Comment:
prefer control structures when available. `for` loop allows `continue` e.g.,
```suggestion
      if (!pendingOffers) continue;
      for (const [_, o] of pendingOffers) {
```

Reply:
Done